examples/graphics/shader-graph-chunks-builder.html

```diff
@@ -0,0 +1,137 @@
+    <script src="../../build/playcanvas.js"></script>
+    <script src="../../build/playcanvas-extras.js"></script>
+    <style>
+        body { 
+            margin: 0;
+            overflow-y: hidden;
+        }
+    </style>
+</head>
+
+<body>
+    <!-- The canvas element -->
+    <canvas id="application-canvas"></canvas>
+    
+    <!-- The script -->
+    <script>
+        var entity, light, camera;
+        var time = 0;
+
+        var canvas = document.getElementById("application-canvas");
+
+        // Create the application and start the update loop
+        var app = new pc.Application(canvas);
+
+        // Set the canvas to fill the window and automatically change resolution to be the same as the canvas size
+        app.setCanvasFillMode(pc.FILLMODE_FILL_WINDOW);
+        app.setCanvasResolution(pc.RESOLUTION_AUTO);
+
+        window.addEventListener("resize", function () {
+            app.resizeCanvas(canvas.width, canvas.height);
+        });
+
+        var miniStats = new pcx.MiniStats(app);
+
+        app.scene.ambientLight = new pc.Color(0.2, 0.2, 0.2);
+
+        // Create an Entity with a camera component
+        camera = new pc.Entity();
+        camera.addComponent("camera", {
+            clearColor: new pc.Color(0.4, 0.45, 0.5)
+        });
+        camera.translate(0, 7, 25);
+        app.root.addChild(camera);
+
+        // Create an entity with a directional light component
+        var light = new pc.Entity();
+        light.addComponent("light", {
+            type: "directional",
+            color: new pc.Color(1, 0.8, 0.25)
+        });
+        app.root.addChild(light);
+        light.setLocalEulerAngles(15, -20, 0);
+
+        var url = "../assets/models/statue.glb";
+        app.assets.loadFromUrl(url, "container", function (err, asset) {
+            entity = new pc.Entity();
+            var modelComponent = entity.addComponent("model", {
+                    type: "asset",
+                    asset: asset.resource.model
+                });
+            app.root.addChild(entity);
+            var model = modelComponent.model;
+            var diffuseMap = model.getMaterials()[0].diffuseMap;
+
+            app.assets.loadFromUrl("../assets/textures/clouds.jpg", "texture", function (err, asset) {    
+                var emissiveMap = asset.resource;
+                app.assets.loadFromUrl("../assets/core_nodes.json", "text", function (err, asset) {
+                    var coreNodeList = asset.resource;
+                    
+                    //register loaded core nodes
+                    pc.shaderChunks.registerCoreShaderGraphNodes(coreNodeList);
+
+                    //create a shader graph builder
+                    var builder = new pc.ShaderGraphBuilder();
+
+                    //add input parameters
+                    var timeParam = builder.addParam('float', 'uTime', 0.0);
+                    var emissiveTexParam = builder.addParam('sampler2D', 'uEmissiveMap', emissiveMap);
```

Comment:
it seems the value is not optional parameter for .addParam - internally in ShaderGraphNode._addIoPort a type of it is tested and used. Do we need to be able to build a shader graph without having a value set?

Reply:
I tried removing the emissiveMap parameter and setting it later and that seems to work. If that's supported, maybe the parameter needs to be marked as optional.

Reply:
yes - it optional - but won't be unless its set later (before rendering) - I'll mark it as optional

Reply:
so should ShaderGraphNode._addIoPort use _type_ instead of _value_ in those ifs? Without value it falls back to what comment suggests is output port.

Reply:
I will adjust the comment

Reply:
done - made value optional when graph params (node ports) are added - and updated comment - thanks!